python_modules/dagster/dagster_tests/definitions_tests/test_computation.py

```diff
@@ -249,3 +252,51 @@ def b() -> str:
         run_config=dg.RunConfig(ops={"my_fn": {"config": {"a": 1, "b": "b"}}}),
     )
     assert result.success
+
+
+@pytest.mark.parametrize("valid", [True, False])
+def test_computation_return_type_value_valid_simple(valid: bool) -> None:
+    class Foo:
+        pass
+
+    @computation(
+        effects=[
+            dg.AssetSpec(key="asset1"),
+            dg.AssetSpec(key="asset2"),
+        ]
+    )
+    def _fn() -> Iterator[dg.MaterializeResult[int]]:
+        yield dg.MaterializeResult(asset_key="asset1", value=1)
+        yield dg.MaterializeResult(asset_key="asset2", value=2 if valid else Foo())  # type: ignore
+
+    assert _fn.node_def.output_dict["asset1"].dagster_type.display_name == "Int"
+    assert _fn.node_def.output_dict["asset2"].dagster_type.display_name == "Int"
+
+    result = dg.materialize_to_memory([_fn.to_assets_def()], raise_on_error=False)
+    assert result.success == valid
+
+
+@pytest.mark.parametrize("valid", [True, False])
+def test_computation_return_type_value_valid_complex(valid: bool) -> None:
+    class Foo:
+        pass
+
+    @computation(
+        effects=[
+            dg.AssetSpec(key="asset1"),
+            dg.AssetSpec(key="asset2"),
+            dg.AssetCheckSpec(name="asset_check1", asset="asset2"),
+        ]
+    )
+    def _fn() -> Iterator[
+        Union[dg.MaterializeResult[int], dg.AssetCheckResult, dg.MaterializeResult[str]]
+    ]:
+        yield dg.MaterializeResult(asset_key="asset1", value=1)
+        yield dg.MaterializeResult(asset_key="asset2", value="a" if valid else Foo())  # type: ignore
+        yield dg.AssetCheckResult(asset_key="asset2", passed=True)
```

Comment:
Yeah I don't really get this one given that can't map the types to specific asset keys